### PR TITLE
fix: Panic/correctness issue in variance GroupsAccumulator

### LINKED ISSUE #12616: Panic/correctness issue in variance GroupsAccumulator
### Describe the bug

When using the sample variance and there is no values in a group we will do a `-1` on a 0 usize. This leads to wrong results in release mode and panic in debug.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by eejbyfeldt**
## Which issue does this PR close?

Closes #12616.

## Rationale for this change

The bug was in the orginal implementation in #12095. This fixes the
issue and modify a test case such that it would have caught it.

Current code panics in debug builds and produces incorrect results in release.

## What changes are included in this PR?
Use `staturating_sub` instead to correct the behavior.

## Are these changes tested?
Modified test case.

## Are there any user-facing changes?
No,

### CODE REVIEW

File: datafusion/functions-aggregate/src/variance.rs [hunk 1]
```diff
@@ -470,7 +470,7 @@ impl VarianceGroupsAccumulator {
 
         if let StatsType::Sample = self.stats_type {
             counts.iter_mut().for_each(|count| {
-                *count -= 1;
+                *count = count.saturating_sub(1);
```
Comment (jayzhan211) [inline datafusion/functions-aggregate/src/variance.rs:473]: Should we return error if count is already 0?

Comment (eejbyfeldt) [reply datafusion/functions-aggregate/src/variance.rs:473]: No, I think producing null is the expected behavior. Note that we use the counts to calculate the null array after this.

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by alamb** [comment]
Thanks @eejbyfeldt and @jayzhan211

**Pull request merged**